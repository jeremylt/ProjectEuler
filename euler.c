#include <stdio.h>
#include "euler.h"
#include "utils.h"

// -----------------------------------------------------------------------------
// Driver
// -----------------------------------------------------------------------------
int main(void) {
  int menuOption = 0;

  printf("-------------------------------------------------------------------\n"
         "  Welcome to a Project Euler solver:\n"
         "-------------------------------------------------------------------\n"
         "  Menu Option 1: Print help information\n"
         "  Menu Option 2: Solve Euler problem\n"
         "  Menu Option 3: Exit\n\n");

  // Loop until exit
  while (menuOption != 3) {
    // Prompt user
    getValidUserInt("  Enter a menu option (1-3): ", stdin, 1, 3, &menuOption);

    // Take selected action
    switch (menuOption) {
    case 1:
      printf("---------------------------------------------------------------\n"
             "  Help:\n"
             "    Menu Option 1: Print help information\n"
             "    Menu Option 2: Solve Euler problem\n"
             "    Menu Option 3: Exit\n\n");
      break;

    case 2: {
      int problemOption;

      printf("---------------------------------------------------------------\n"
             "  Euler Problems 1-%d are avaliable\n", NUM_PROBLEMS);
      getValidUserInt("  Select an Euler problem to solve: ", stdin, 1,
                      NUM_PROBLEMS, &problemOption);

      (*problems[problemOption-1])(stdin);
    } break;

    default :
      printf("---------------------------------------------------------------\n"
             "  Exit\n\n"
             "---------------------------------------------------------------\n"
            );
      break;
    }
  }

  return 0;
}
// -----------------------------------------------------------------------------
