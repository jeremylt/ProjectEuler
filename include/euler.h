#ifndef EULER_H
#define EULER_H
#include <math.h>

#define NUM_PROBLEMS 7

// -----------------------------------------------------------------------------
// Solution code
// -----------------------------------------------------------------------------
int problem001(long a, long b, long limit, long *sum);
int problem002(long a, long limit, long *sum);
int problem003(long number, long *factor);
int problem004(int digits, long *palindrome);
int problem005(int limit, long *lcm);
int problem006(int limit, long *difference);
int problem007(int n, long *prime);

// -----------------------------------------------------------------------------
// Command line interfaces
// -----------------------------------------------------------------------------
typedef int funcPtr(FILE *stream);
funcPtr problem001CL, problem002CL, problem003CL, problem004CL, problem005CL,
        problem006CL, problem007CL;

funcPtr *problems[NUM_PROBLEMS] = {
  problem001CL,
  problem002CL,
  problem003CL,
  problem004CL,
  problem005CL,
  problem006CL,
  problem007CL
};

#endif
