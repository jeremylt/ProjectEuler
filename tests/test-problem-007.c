/// @file
/// Test Project Euler Problem 7
/// \test Test problem 7 result
#include <stdio.h>
#include "euler.h"
// -----------------------------------------------------------------------------
int main(void) {
  FILE *stream;

  // Open file
  stream = fopen("tests/input/test-problem-007.in", "r");
  if(stream == NULL) {
    // LCOV_EXCL_START
    perror("Error opening file\n");
    return(-1);
    // LCOV_EXCL_STOP
  }

  // Get 6th prime
  problem007CL(stream);

  // Get 10001th prime
  problem007CL(stream);

  // Close
  fclose(stream);

  return 0;
}
// -----------------------------------------------------------------------------
