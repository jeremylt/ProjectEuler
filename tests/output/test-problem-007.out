    This code finds the [n]th prime.

      Enter n:     The 6th prime is: 13

    This code finds the [n]th prime.

      Enter n:     The 10001th prime is: 104743

